### PR TITLE
client.go: fix race in Register & use RWMutex

### PR DESCRIPTION
**Pull request by unclejack**
This PR fixes one remaining race condition in register and switches to RWMutex.

RWMutex is used by https://github.com/socketplane/libovsdb/commit/c39b98174278cfca364a81d8f5bca2af0a423db2#diff-d0b7f86cea13f90d3cc0cc2338da8d5fR28.

That upstream commit didn't fix the race in Register.
It also makes a small change to remove a redundant select.

### CODE REVIEW

**Review by dseevr — APPROVED** [review]

File: client.go [hunk 1]
```diff
@@ -252,16 +254,14 @@ func getTableUpdatesFromRawUnmarshal(raw map[string]map[string]RowUpdate) TableU
...
 func clearConnection(c *rpc2.Client) {
 	connectionsMutex.Lock()
+	defer connectionsMutex.Unlock()
 	delete(connections, c)
-	connectionsMutex.Unlock()
 }
 
 func handleDisconnectNotification(c *rpc2.Client) {
 	disconnected := c.DisconnectNotify()
-	select {
```
Comment (unclejack) [inline client.go:261]: This select was redundant. select makes sense when there are multiple possible cases. It  doesn't make sense to use select when there's a single possible case.

This changes the code to read once from the channel and perform that operation. The effect is the same.

File: client.go [hunk 2]
```diff
@@ -252,16 +254,14 @@ func getTableUpdatesFromRawUnmarshal(raw map[string]map[string]RowUpdate) TableU
 
 func clearConnection(c *rpc2.Client) {
 	connectionsMutex.Lock()
+	defer connectionsMutex.Unlock()
```
Comment (unclejack) [inline client.go:257]: This is needed to ensure the code actually executes.

File: client.go [hunk 3]
```diff
@@ -145,13 +147,13 @@ func update(client *rpc2.Client, params []interface{}, reply *interface{}) error
 
 	// Update the local DB cache with the tableUpdates
 	tableUpdates := getTableUpdatesFromRawUnmarshal(rowUpdates)
-	connectionsMutex.Lock()
+	connectionsMutex.RLock()
```
Comment (unclejack) [inline client.go:150]: We're only reading here. We should only lock for reading here.

File: client.go [hunk 4]
```diff
@@ -20,15 +20,15 @@ type OvsdbClient struct {
...
+	defer connectionsMutex.Unlock()
 	connections[c] = ovs
-	connectionsMutex.Unlock()
 	return ovs
 }
 
 // Would rather replace this connection map with an OvsdbClient Receiver scoped method
 // Unfortunately rpc2 package acts wierd with a receiver scoped method and needs some investigation.
 var connections map[*rpc2.Client]*OvsdbClient = make(map[*rpc2.Client]*OvsdbClient)
-var connectionsMutex sync.Mutex
```
Comment (unclejack) [inline client.go:31]: This is changed to use a RWMutex. It's a better idea to do this.

File: client.go [hunk 5]
```diff
@@ -90,6 +90,8 @@ func Connect(ipAddr string, port int) (*OvsdbClient, error) {
 }
 
 func (ovs *OvsdbClient) Register(handler NotificationHandler) {
+	connectionsMutex.Lock()
```
Comment (unclejack) [inline client.go:93]: This is needed to fix a race.

**Comment by unclejack** [comment]
@jojimt @abhinandanpb: Please take a look at this PR. This is needed to fix bugs in libovsdb. I've commented to explain why the changes were made. Some changes are only commented in one place because they make the same change.

This fix is needed.

**Comment by jojimt** [comment]
LGTM

**Pull request merged**